libvips/foreign/dcrawload.c

```diff
@@ -227,25 +356,98 @@ vips_foreign_load_dcraw_class_init(VipsForeignLoadDcRawClass *class)
 	GObjectClass *gobject_class = G_OBJECT_CLASS(class);
 	VipsObjectClass *object_class = (VipsObjectClass *) class;
 	VipsOperationClass *operation_class = VIPS_OPERATION_CLASS(class);
+	VipsForeignClass *foreign_class = (VipsForeignClass *) class;
 	VipsForeignLoadClass *load_class = (VipsForeignLoadClass *) class;
 
 	gobject_class->dispose = vips_foreign_load_dcraw_dispose;
+	gobject_class->set_property = vips_object_set_property;
+	gobject_class->get_property = vips_object_get_property;
 
 	object_class->nickname = "dcrawload_base";
 	object_class->description = _("load RAW camera files");
-	object_class->build = vips_foreign_load_dcraw_build;
 
-	/* LibRaw is fuzzed, but not by us.
-	 */
 	operation_class->flags |= VIPS_OPERATION_UNTRUSTED;
 
+	/* We need to be ahead of JPEG and TIFF, since many cameras use those
+	 * formats as containers. We are very slow to open, but we only test the
+	 * filename suffix, so that's fine.
+	 */
+	foreign_class->priority = 100;
+
 	load_class->get_flags = vips_foreign_load_dcraw_get_flags;
 	load_class->header = vips_foreign_load_dcraw_header;
-	load_class->load = NULL;
+	load_class->load = vips_foreign_load_dcraw_load;
+
+	VIPS_ARG_INT(class, "bitdepth", 12,
+		_("Bit depth"),
+		_("Number of bits per pixel"),
+		VIPS_ARGUMENT_OPTIONAL_INPUT,
+		G_STRUCT_OFFSET(VipsForeignLoadDcRaw, bitdepth),
+		8, 16, 8);
+
 }
 
 static void
 vips_foreign_load_dcraw_init(VipsForeignLoadDcRaw *raw)
+{
+	raw->bitdepth = 8;
+}
+
+typedef struct _VipsForeignLoadDcRawSource {
+	VipsForeignLoadDcRaw parent_object;
+
+	/* Load from a source.
+	 */
+	VipsSource *source;
+
+} VipsForeignLoadDcRawSource;
+
+typedef VipsForeignLoadDcRawClass VipsForeignLoadDcRawSourceClass;
+
+G_DEFINE_TYPE(VipsForeignLoadDcRawSource, vips_foreign_load_dcraw_source,
+	vips_foreign_load_dcraw_get_type());
+
+static int
+vips_foreign_load_dcraw_source_build(VipsObject *object)
+{
+	VipsForeignLoadDcRaw *dcraw = (VipsForeignLoadDcRaw *) object;
+	VipsForeignLoadDcRawSource *source = (VipsForeignLoadDcRawSource *) object;
+
+	if (source->source) {
+		dcraw->source = source->source;
+		g_object_ref(dcraw->source);
+	}
+
+	return VIPS_OBJECT_CLASS(vips_foreign_load_dcraw_source_parent_class)
+		->build(object);
+}
+
+static void
+vips_foreign_load_dcraw_source_class_init(
+	VipsForeignLoadDcRawSourceClass *class)
+{
+	GObjectClass *gobject_class = G_OBJECT_CLASS(class);
+	VipsObjectClass *object_class = (VipsObjectClass *) class;
+	VipsOperationClass *operation_class = VIPS_OPERATION_CLASS(class);
+
+	gobject_class->set_property = vips_object_set_property;
+	gobject_class->get_property = vips_object_get_property;
+
+	object_class->nickname = "dcrawload_source";
+	object_class->build = vips_foreign_load_dcraw_source_build;
+
```

Comment:
Perhaps we should also a `->is_a_source` that always returns `FALSE`? Similar to `csvload_source`:
https://github.com/libvips/libvips/blob/8c5c65fddc782dc5914caf52ea4c17de61f9cb73/libvips/foreign/csvload.c#L625-L633

Otherwise, `vips_foreign_find_load_source()` would trigger this warning:
https://github.com/libvips/libvips/blob/8c5c65fddc782dc5914caf52ea4c17de61f9cb73/libvips/foreign/foreign.c#L751-L752

Reply:
Oh, good point!